SSWC/doc/man/man3/SpaceBody.3
.TH "SpaceBody" 3 "Mon May 9 2016" "Version 0.1" "MissionsVisualizer" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SpaceBody \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <SpaceBody\&.h>\fP
.PP
Inherits \fBSpaceObject\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBBulkParameter\fP { \fBBP_RADIUS\fP, \fBBP_GM\fP, \fBBP_MASS\fP, \fBBP_ACC\fP, \fBBP_POLE_RA\fP, \fBBP_POLE_DEC\fP, \fBBP_PM\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSpaceBody\fP (long \fBspiceId\fP, const \fBstd::string\fP &\fBname\fP='')"
.br
.ti -1c
.RI "\fBSpaceBody\fP (const \fBstd::string\fP &spiceName, const \fBstd::string\fP &\fBname\fP='')"
.br
.ti -1c
.RI "\fBSpaceBody\fP (const \fBSpaceObject\fP &\fBobj\fP)"
.br
.ti -1c
.RI "\fB~SpaceBody\fP ()"
.br
.ti -1c
.RI "virtual \fBSpaceBody\fP * \fBClone\fP () const "
.br
.ti -1c
.RI "bool \fBHasIAUFrame\fP () const "
.br
.ti -1c
.RI "\fBFrame\fP \fBGetIAUFrame\fP () const "
.br
.ti -1c
.RI "bool \fBHasDefaultFrame\fP () const "
.br
.ti -1c
.RI "\fBFrame\fP \fBGetDefaultFrame\fP () const "
.br
.ti -1c
.RI "bool \fBHasParameter\fP (\fBBulkParameter\fP \fBparam\fP) const "
.br
.ti -1c
.RI "double \fBGetSingleDimParam\fP (\fBBulkParameter\fP \fBparam\fP) const "
.br
.ti -1c
.RI "std::vector< double > \fBGetMultiDimParam\fP (\fBBulkParameter\fP \fBparam\fP) const "
.br
.ti -1c
.RI "\fBLength\fP \fBGetRadius\fP () const "
.br
.ti -1c
.RI "\fBstd::array\fP< \fBLength\fP, 3 > \fBGetRadii\fP () const "
.br
.ti -1c
.RI "\fBGravitationalParameter\fP \fBGetGM\fP () const "
.br
.ti -1c
.RI "\fBMass\fP \fBGetMass\fP () const "
.br
.ti -1c
.RI "\fBAcceleration\fP \fBGetSurfaceAcceleration\fP () const "
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
Definition at line 11 of file SpaceBody\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBSpaceBody::BulkParameter\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIBP_RADIUS \fP\fP
.TP
\fB\fIBP_GM \fP\fP
.TP
\fB\fIBP_MASS \fP\fP
.TP
\fB\fIBP_ACC \fP\fP
.TP
\fB\fIBP_POLE_RA \fP\fP
.TP
\fB\fIBP_POLE_DEC \fP\fP
.TP
\fB\fIBP_PM \fP\fP
.PP
Definition at line 14 of file SpaceBody\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SpaceBody::SpaceBody (longspiceId, const \fBstd::string\fP &name = \fC''\fP)"

.PP
Definition at line 3 of file SpaceBody\&.cpp\&.
.SS "SpaceBody::SpaceBody (const \fBstd::string\fP &spiceName, const \fBstd::string\fP &name = \fC''\fP)"

.PP
Definition at line 8 of file SpaceBody\&.cpp\&.
.SS "SpaceBody::SpaceBody (const \fBSpaceObject\fP &obj)"

.PP
Definition at line 13 of file SpaceBody\&.cpp\&.
.SS "SpaceBody::~SpaceBody ()"

.PP
Definition at line 27 of file SpaceBody\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBSpaceBody\fP * SpaceBody::Clone () const\fC [virtual]\fP"

.PP
Reimplemented from \fBSpaceObject\fP\&.
.PP
Definition at line 36 of file SpaceBody\&.cpp\&.
.SS "\fBFrame\fP SpaceBody::GetDefaultFrame () const"

.PP
Definition at line 69 of file SpaceBody\&.cpp\&.
.SS "\fBGravitationalParameter\fP SpaceBody::GetGM () const"

.PP
Definition at line 216 of file SpaceBody\&.cpp\&.
.SS "\fBFrame\fP SpaceBody::GetIAUFrame () const"

.PP
Definition at line 51 of file SpaceBody\&.cpp\&.
.SS "\fBMass\fP SpaceBody::GetMass () const"

.PP
Definition at line 223 of file SpaceBody\&.cpp\&.
.SS "std::vector< double > SpaceBody::GetMultiDimParam (\fBBulkParameter\fPparam) const"

.PP
Definition at line 155 of file SpaceBody\&.cpp\&.
.SS "\fBstd::array\fP< \fBLength\fP, 3 > SpaceBody::GetRadii () const"

.PP
Definition at line 204 of file SpaceBody\&.cpp\&.
.SS "\fBLength\fP SpaceBody::GetRadius () const"

.PP
Definition at line 197 of file SpaceBody\&.cpp\&.
.SS "double SpaceBody::GetSingleDimParam (\fBBulkParameter\fPparam) const"

.PP
Definition at line 119 of file SpaceBody\&.cpp\&.
.SS "\fBAcceleration\fP SpaceBody::GetSurfaceAcceleration () const"

.PP
Definition at line 230 of file SpaceBody\&.cpp\&.
.SS "bool SpaceBody::HasDefaultFrame () const"

.PP
Definition at line 58 of file SpaceBody\&.cpp\&.
.SS "bool SpaceBody::HasIAUFrame () const"

.PP
Definition at line 41 of file SpaceBody\&.cpp\&.
.SS "bool SpaceBody::HasParameter (\fBBulkParameter\fPparam) const"

.PP
Definition at line 83 of file SpaceBody\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for MissionsVisualizer from the source code\&.


SSWC/src/CSpice/SpaceBody.cpp
#include "SpaceBody.h"

SpaceBody::SpaceBody(long spiceId, const std::string& name) : SpaceObject(spiceId, name)
{
	Init();
}

SpaceBody::SpaceBody(const std::string& spiceName, const std::string& name) : SpaceObject(spiceName, name)
{
	Init();
}

SpaceBody::SpaceBody(const SpaceObject& obj) : SpaceObject(obj.GetSpiceId(), obj.GetName())
{
	Init();
}

void SpaceBody::Init()
{
	//frame = nullptr;
	//if(HasIAUFrame())
	//{
	//	UseBodyFrame(GetIAUFrame());
	//}
}

SpaceBody::~SpaceBody()
{
	//if(this->frame != nullptr)
	//{
	//	delete this->frame;
	//	this->frame = nullptr;
	//}
}

SpaceBody* SpaceBody::Clone() const
{
	return new SpaceBody(*this);
}

bool SpaceBody::HasIAUFrame() const
{
	std::string frameName = "IAU_" + GetSpiceName();

	long frcode;
	CSPICE_ASSERT( namfrm_c(frameName.c_str(), &frcode) );

	return frcode != 0;
}

Frame SpaceBody::GetIAUFrame() const
{
	std::string frameName = "IAU_" + GetSpiceName();

	return Frame(frameName);
}

bool SpaceBody::HasDefaultFrame() const
{
	char frameName[FRAME_NAME_MAX_LENGTH];
	long frameId;
	SpiceBoolean found;

	CSPICE_ASSERT(cidfrm_c(spiceId, FRAME_NAME_MAX_LENGTH, &frameId, frameName, &found));

	return found != SPICEFALSE;
}

Frame SpaceBody::GetDefaultFrame() const
{
	char frameName[FRAME_NAME_MAX_LENGTH];
	long frameId;
	SpiceBoolean found;

	CSPICE_ASSERT(cidfrm_c(spiceId, FRAME_NAME_MAX_LENGTH, &frameId, frameName, &found));

	if(!found)
		CSpiceUtil::SignalError("No such frame available");

	return Frame(frameId);
}

bool SpaceBody::HasParameter(BulkParameter param) const
{
	switch(param)
	{
	case BP_RADIUS:
		return bodfnd_c(spiceId, "RADII") != SPICEFALSE;
		break;

	case BP_GM:
		return bodfnd_c(spiceId, "GM") != SPICEFALSE;
		break;

	case BP_MASS:
		return HasParameter(BP_GM);
		break;

	case BP_ACC:
		return HasParameter(BP_GM) && HasParameter(BP_RADIUS);
		break;

	case BP_POLE_RA:
		return	(bodfnd_c(spiceId, "POLE_RA") != SPICEFALSE);
		break;

	case BP_POLE_DEC:
		return	(bodfnd_c(spiceId, "POLE_DEC") != SPICEFALSE);
		break;

	case BP_PM:
		return	(bodfnd_c(spiceId, "PM") != SPICEFALSE);
		break;
	}

	return false;
}

double SpaceBody::GetSingleDimParam(BulkParameter param) const
{
	double value = -1.0;
	long dim;

	std::vector<double> radii;
	double GM;
	double radius;

	switch(param)
	{
	case BP_RADIUS:
		radii = GetMultiDimParam(BP_RADIUS);
		value = std::pow(radii[0] * radii[1] * radii[2], 1.0 / 3.0); // volumetric mean
		break;

	case BP_GM:
		CSPICE_ASSERT( bodvcd_c(spiceId, "GM", 1, &dim, &value) );
		//value *= 1000.0 * 1000.0 * 1000.0; // km^3 -> m^3
		break;

	case BP_MASS:
		GM = GetSingleDimParam(BP_GM);
		value = GM / G;
		break;

	case BP_ACC:
		GM = GetSingleDimParam(BP_GM);
		radius = GetSingleDimParam(BP_RADIUS);
		value = GM / (radius * radius);
		break;
	}

	return value;
}

std::vector<double> SpaceBody::GetMultiDimParam(BulkParameter param) const
{
	std::vector<double> values;
	long dim;

	switch(param)
	{
	case BP_RADIUS:
		double radii[3];
		CSPICE_ASSERT( bodvcd_c(spiceId, "RADII", 3, &dim, radii) );
		//for(int i = 0; i < 3; i++)
		//	radii[i] *= 1000.0; // km -> m
		for(int i = 0; i < 3; i++)
			values.push_back(radii[i]);
		break;

	case BP_POLE_RA:
		double pole_ra[3];
		CSPICE_ASSERT(bodvcd_c(spiceId, "POLE_RA", 3, &dim, pole_ra));
		for(int i = 0; i < 3; i++)
			values.push_back(pole_ra[i]);
		break;

	case BP_POLE_DEC:
		double pole_dec[3];
		CSPICE_ASSERT(bodvcd_c(spiceId, "POLE_DEC", 3, &dim, pole_dec));
		for(int i = 0; i < 3; i++)
			values.push_back(pole_dec[i]);
		break;

	case BP_PM:
		double pm[3];
		CSPICE_ASSERT(bodvcd_c(spiceId, "PM", 3, &dim, pm));
		for(int i = 0; i < 3; i++)
			values.push_back(pm[i]);

		break;
	}

	return values;
}

Length SpaceBody::GetRadius() const
{
	double value = GetSingleDimParam(BP_RADIUS);

	return Length(value, Units::Metric::kilometers);
}

std::array<Length, 3> SpaceBody::GetRadii() const
{
	std::array<Length, 3> radii;
	std::vector<double> values = GetMultiDimParam(BP_RADIUS);
	for(int i = 0; i < 3; i++)
	{
		radii[i] = Length(values[i], Units::Metric::kilometers);
	}

	return radii;
}

GravitationalParameter SpaceBody::GetGM() const
{
	double value = GetSingleDimParam(BP_GM);

	return GravitationalParameter(value, Units::Metric::gm);
}

Mass SpaceBody::GetMass() const
{
	double value = GetSingleDimParam(BP_MASS);

	return Mass(value, Units::Metric::kilograms);
}

Acceleration SpaceBody::GetSurfaceAcceleration() const
{
	double value = GetSingleDimParam(BP_ACC);

	return Acceleration(value, AccelerationUnit::DerivedUnit(Units::Metric::kilometers, Units::Common::seconds));
}
